### PR TITLE
Python 3.12.0rc2

### PR DESCRIPTION
**Pull request by xhochy**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/patches/0017-Fix-TZPATH-on-windows.patch
Comment (isuruf) [inline recipe/patches/0017-Fix-TZPATH-on-windows.patch:2]: ?

File: recipe/patches/0022-Unvendor-tcltk.patch
Comment (isuruf) [inline recipe/patches/0022-Unvendor-tcltk.patch:2]: ?

File: recipe/patches/0017-Fix-TZPATH-on-windows.patch
Comment (xhochy) [reply recipe/patches/0017-Fix-TZPATH-on-windows.patch:2]: My start was to copy over the patches from Anaconda's 3.12 effort. It seems like they have changed the attribution on some of the patches. I can revert this here later.

**Comment by xhochy** [comment]
@isuruf @erykoff This is running into https://bugs.python.org/issue44556 again. Do you have any ideas on how to proceed/where I should debug?

**Comment by xhochy** [comment]
> @isuruf @erykoff This is running into https://bugs.python.org/issue44556 again. Do you have any ideas on how to proceed/where I should debug?

Found a workaround.

**Review by h-vetinari — COMMENTED** [review]
It think the patches didn't undergo crlf-normalization, see [here](https://github.com/conda-forge/python-feedstock/blob/main/recipe/patches/README.md). Otherwise looks very good already!

**Comment by xhochy** [comment]
@h-vetinari I ran crlf normalization but I'm still unable to `git am` all the patches locally. By using `--keep-cr`, I can apply most of them but fail with `0015-Doing-d1trimfile.patch`:

```
% git am --keep-cr $RECIPE_DIR/recipe/patches/0015-Doing-d1trimfile.patch
warning: quoted CRLF detected
Applying: Doing d1trimfile
error: patch failed: PCbuild/_socket.vcxproj:95
error: PCbuild/_socket.vcxproj: patch does not apply
error: patch failed: PCbuild/_testcapi.vcxproj:92
error: PCbuild/_testcapi.vcxproj: patch does not apply
error: patch failed: PCbuild/pythonw.vcxproj:91
error: PCbuild/pythonw.vcxproj: patch does not apply
Patch failed at 0001 Doing d1trimfile
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

Do you have an idea which flags I need to amend them together so I can get a clear re-export?

File: recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch
Comment (h-vetinari) [inline recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch:13]: This patch now adds an enormous c-implementation of I don't know what exactly

**Review by h-vetinari — COMMENTED** [review]
> Do you have an idea which flags I need to amend them together so I can get a clear re-export?

I fixed up the patches so that they can now be applied on a fresh checkout of the tag as:
```
find ../path/to/feedstock/recipe/patches/ | xargs git am
```
resp.
```
for /f %f in ('dir /b /S ..\path\to\feedstock\recipe\patches') do git am %f
```

I also renamed the patches to keep their filenames from master, so that it'll be easier to review the diff of the patches (instead of the GH UI pretending a sufficiently changed patch was deleted and recreated).

File: recipe/patches/0009-Unvendor-openssl.patch
Comment (h-vetinari) [inline recipe/patches/0009-Unvendor-openssl.patch:180]: We should probably simply go for outright deletion here, rather than commenting this out.

File: recipe/patches/0015-cross-compile-darwin.patch
Comment (h-vetinari) [inline recipe/patches/0015-cross-compile-darwin.patch:28]: AFAICT, this hunk would still apply. Do we not need it anymore?

File: recipe/patches/0021-Unvendor-tcltk.patch
Comment (h-vetinari) [inline recipe/patches/0021-Unvendor-tcltk.patch:46]: I think we might still want to delete this as before and replace it with 
```
    <tclWin32Exe>$(condaDir)\bin\tclsh$(TclMajorVersion)$(TclMinorVersion)t.exe</tclWin32Exe>
```
?

File: recipe/patches/0025-Unvendor-expat.patch
Comment (h-vetinari) [inline recipe/patches/0025-Unvendor-expat.patch:4]: What's the reason this patch was deleted?

File: recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch
Comment (h-vetinari) [reply recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch:13]: At first sight, it looks like it could be an accidental recommit of [`Modules/getpath.c`](https://github.com/python/cpython/blob/v3.12.0rc1/Modules/getpath.c), but it's not.

After a bit more digging, this file was last [present](https://github.com/python/cpython/blob/v3.10.13/PC/getpathp.c) in 3.10, and was removed by https://github.com/python/cpython/commit/99fcf1505218464c489d419d4500f126b6d6dc28. Not sure why we'd need to bring it back if we didn't need it for 3.11?

**Review by h-vetinari — COMMENTED** [review]
Looking a bit more at some of the patch diffs, I think there are still some clean-ups necessary.

File: recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch
Comment (xhochy) [reply recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch:13]: This could be an artefact of me importing the patches from Anaconda. Maybe we dropped it in conda-forge but not on the Anaconda side. If Python doesn't ship it anymore, I feel like we should try to also not ship it and see if someone reports an issue. If so, we can re-add with a comment explaining the situation.

Comment (h-vetinari) [reply recipe/patches/0008-Disable-registry-lookup-unless-CONDA_PY_ALLOW_REG_PA.patch]: Seems to work fine without. :)

File: recipe/patches/0015-cross-compile-darwin.patch
Comment (h-vetinari) [reply recipe/patches/0015-cross-compile-darwin.patch]: Readded.

File: recipe/patches/0021-Unvendor-tcltk.patch
Comment (h-vetinari) [reply recipe/patches/0021-Unvendor-tcltk.patch]: Done.

File: recipe/patches/0025-Unvendor-expat.patch
Comment (h-vetinari) [reply recipe/patches/0025-Unvendor-expat.patch]: Readded this patch

**Comment by xhochy** [comment]
@conda-forge/python This is ready for review and I would be eager to merge. Personally, I would merge this to main and create a 3.11 branch immediately. I don't see a benefit of using a really short-lived `rc` branch.

File: recipe/build_base.sh
Comment (isuruf) [inline recipe/build_base.sh:409]:
```suggestion
ln -s ${PREFIX}/bin/python${VER} ${PREFIX}/bin/python3.1
```

File: recipe/build_base.sh
Comment (isuruf) [inline recipe/build_base.sh:536]:
```suggestion
ln -s "${PREFIX}/lib/python${VER} "${PREFIX}/lib/python3.1"
```

File: recipe/patches/0027-Don-t-checksharedmods.patch
Comment (isuruf) [inline recipe/patches/0027-Don-t-checksharedmods.patch:4]: Why is this done?

Comment (xhochy) [reply recipe/patches/0027-Don-t-checksharedmods.patch:4]: Running these checks on installations fails with the following. I wouldn't expect them to work in a cross-compiled setting ever. 

```
Traceback (most recent call last):
  File "/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/Tools/build/check_extension_modules.py", line 25, in <module>
    import pathlib
  File "/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/Lib/pathlib.py", line 20, in <module>
    from urllib.parse import quote_from_bytes as urlquote_from_bytes
  File "/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/Lib/urllib/parse.py", line 36, in <module>
    import math
ImportError: dlopen(/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/build/lib.darwin-aarch64-3.12/math.cpython-312-darwin.so, 0x0002): tried: '/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/build/lib.darwin-aarch64-3.12/math.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/build/lib.darwin-aarch64-3.12/math.cpython-312-darwin.so' (no such file), '/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/build/lib.darwin-aarch64-3.12/math.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/Modules/math.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/Modules/math.cpython-312-darwin.so' (no such file), '/Users/uwe/mambaforge-x64/conda-bld/python-split_1694676471721/work/build-static/Modules/math.cpython-312-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
```

Comment (xhochy) [reply recipe/patches/0027-Don-t-checksharedmods.patch:4]: @isuruf Is that OK? Should I add a more detailed comment or do you want me to look into this further (if so: where/what?)?

**Review by ocefpaf — APPROVED** [review]

File: recipe/patches/0027-Don-t-checksharedmods.patch
Comment (isuruf) [reply recipe/patches/0027-Don-t-checksharedmods.patch:4]: Yes, but can you apply the patch for when `build_platform != target_platform and CROSSCOMPILING_EMULATOR == ""` so that the patch is applied only when python is cross compiled?

**Review by h-vetinari — APPROVED** [review]
I don't know all the ins & outs of this feedstock, but FWIW, this LGTM (especially behind an rc-label; we can IMO fix any potentially remaining details for GA).

**Pull request merged**
